netplan/configmanager.py

```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/python3
+#
+# Copyright (C) 2018 Canonical, Ltd.
+# Author: Mathieu Trudel-Lapierre <mathieu.trudel-lapierre@canonical.com>
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; version 3.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+'''netplan configuration manager'''
+
+import logging
+import os
+import shutil
+import sys
+import tempfile
+
+
+class ConfigManager(object):
+
+    def __init__(self, prefix="/"):
+        self.prefix = prefix
+        self.tempdir = tempfile.mkdtemp(prefix='netplan_')
+        self.temp_etc = os.path.join(self.tempdir, "etc")
+        self.temp_run = os.path.join(self.tempdir, "run")
+        self.extra_files = {}
+
+    def add(self, config_dict):
+        for config_file in config_dict:
+            self._copy_file(config_file, config_dict[config_file])
+        self.extra_files.update(config_dict)
+
+    def backup(self, backup_config_dir=True):
+        if backup_config_dir:
+            self._copy_tree(os.path.join(self.prefix, "etc/netplan"),
+                            os.path.join(self.temp_etc, "netplan"))
+        self._copy_tree(os.path.join(self.prefix, "run/NetworkManager/system-connections"),
+                        os.path.join(self.temp_run, "NetworkManager", "system-connections"),
+                        missing_ok=True)
+        self._copy_tree(os.path.join(self.prefix, "run/systemd/network"),
+                        os.path.join(self.temp_run, "systemd", "network"),
+                        missing_ok=True)
+
+    def revert(self):
+        try:
+            for extra_file in self.extra_files.values():
+                os.unlink(extra_file)
+            temp_nm_path = "{}/NetworkManager/system-connections".format(self.temp_run)
+            temp_networkd_path = "{}/systemd/network".format(self.temp_run)
+            if os.path.exists(temp_nm_path):
+                shutil.rmtree(os.path.join(self.prefix, "run/NetworkManager/system-connections"))
+                self._copy_tree(temp_nm_path,
+                                os.path.join(self.prefix, "run/NetworkManager/system-connections"))
+            if os.path.exists(temp_networkd_path):
+                shutil.rmtree(os.path.join(self.prefix, "run/systemd/network"))
+                self._copy_tree(temp_networkd_path,
+                                os.path.join(self.prefix, "run/systemd/network"))
```

Comment:
The virtual interface may be left behind, but the previous config will be re-applied -- there's no harm in leaving the interface behind, aside from it being unnecessary.

As for potentially un-revertable changes (bridge params), I agree we should exit non-zero instead of trying at all.

I'll write something to skip testing at all in the case of those parameters, but there are still corner cases that can't be covered until we can also diff config (which will be a separate feature on its own)